src/compiler/visitorPublic.ts

```diff
@@ -378,929 +378,799 @@ namespace ts {
     export function visitEachChild<T extends Node>(node: T | undefined, visitor: Visitor, context: TransformationContext, nodesVisitor?: typeof visitNodes, tokenVisitor?: Visitor): T | undefined;
     /* @internal */
     export function visitEachChild<T extends Node>(node: T | undefined, visitor: Visitor, context: TransformationContext, nodesVisitor?: NodesVisitor, tokenVisitor?: Visitor, nodeVisitor?: NodeVisitor): T | undefined; // eslint-disable-line @typescript-eslint/unified-signatures
-    export function visitEachChild(node: Node | undefined, visitor: Visitor, context: TransformationContext, nodesVisitor = visitNodes, tokenVisitor?: Visitor, nodeVisitor: NodeVisitor = visitNode): Node | undefined {
+    export function visitEachChild<T extends Node>(node: T | undefined, visitor: Visitor, context: TransformationContext, nodesVisitor = visitNodes, tokenVisitor?: Visitor, nodeVisitor: NodeVisitor = visitNode): T | undefined {

```

Comment:
Isn't the input possibly unrelated to the output? @jakebailey noticed something like this while converting to SFT.

Reply:
It totally is. In this case, the implementation fails under SFT because it doesn't overlap with its overloads, except that we don't properly check that without SFT in this case.

Changing the signature to match the overload is totally fine; no worse than what we have now and users won't see the difference. Actual SFT support and a corrected signature will have to come later when we're ready for a break.

Reply:
I'm asserting `visitEachChildTable[node.kind] as VisitEachChildFunction<T> | undefined`. It's difficult to prove this in the type system, but `VisitEachChildTable[T["kind"]]` *should* be `VisitEachChildFunction<T> | undefined`. This isn't perfect, but its better than `any`.

Reply:
We also don't have a great way of expressing the existing overloads in the type system. `visitEachChild<BinaryExpressionButWithExtraTypeInfo>(...)` would say it produces a `BinaryExpressionButWithExtraTypeInfo`, but would actually produce a `BinaryExpressionButWithExtraTypeInfo | BinaryExpression` depending on whether we updated the node. Unless we introduce a big ADT union of _all_ possible `Node` subtypes, I don't think this can be improved.

Reply:
Hm, unless I'm mistaken, the transform API is used heavily to completely and entirely replace nodes when visiting with... whatever, even nodes of differing types. So, I would expect that a fixed signature would instead have an overload with a more specific return first, but also a more generic "it's just a node" when the result can't be inferred.

But, that's offtopic.

Reply:
`visitEachChild` *always* returns the same type of node passed as the first argument. You can only update *child* nodes using the supplied `visitor`.

Reply:
Ah, oops, I think Daniel and I are thinking of `visitNode` and `visitNodes`, which do have a bogus signature.

Reply:
It is and it isn't. `visitNode` is really supposed to be used with a `test` parameter, such that you validate the result is a `T` when visiting a class of nodes such as `Statement`, `Expression`, etc. The `test` is optional, but really shouldn't be. Unfortunately, dropping `test` would probably be a breaking change, and we'd have to expose some of the higher-level node tests (i.e., `isStatement`, `isExpression`, `isClassElement`, etc.) in the public API that are currently internal.